sdk/freeport/freeport.go

```diff
@@ -31,31 +32,131 @@ const (
 )
 
 var (
+	// effectiveMaxBlocks is the number of available port blocks.
+	// lowPort + effectiveMaxBlocks * blockSize must be less than 65535.
+	effectiveMaxBlocks int
+
 	// firstPort is the first port of the allocated block.
 	firstPort int
 
 	// lockLn is the system-wide mutex for the port block.
 	lockLn net.Listener
 
-	// mu guards nextPort
+	// mu guards the mutable data
 	mu sync.Mutex
 
 	// once is used to do the initialization on the first call to retrieve free
 	// ports
 	once sync.Once
 
-	// port is the last allocated port.
-	port int
+	// condNotEmpty is a condition variable to wait for freePorts to be not
+	// empty
+	condNotEmpty *sync.Cond
+
+	// freePorts is a FIFO of all currently free ports. Take from the front,
+	// and return to the back.
+	freePorts *list.List
+
+	// pendingPorts is a FIFO of recently freed ports that have not yet passed
+	// the not-in-use check.
+	pendingPorts *list.List
+
+	// total is the total number of available ports in the block for use.
+	total int
 )
 
 // initialize is used to initialize freeport.
 func initialize() {
-	if lowPort+maxBlocks*blockSize > 65535 {
+	effectiveMaxBlocks = adjustMaxBlocks()
+	if effectiveMaxBlocks < 0 {
+		panic("freeport: no blocks of ports available outside of ephemeral range")
+	}
+	if lowPort+effectiveMaxBlocks*blockSize > 65535 {
 		panic("freeport: block size too big or too many blocks requested")
 	}
 
 	rand.Seed(time.Now().UnixNano())
 	firstPort, lockLn = alloc()
+
+	condNotEmpty = sync.NewCond(&mu)
+	freePorts = list.New()
+	pendingPorts = list.New()
+
+	// fill with all available free ports
+	for port := firstPort + 1; port < firstPort+blockSize; port++ {
+		if used := isPortInUse(port); !used {
+			freePorts.PushBack(port)
+		}
+	}
+	total = freePorts.Len()
+
+	go checkFreedPorts()
+}
+
+func checkFreedPorts() {
+	ticker := time.NewTicker(250 * time.Millisecond)
+	for {
+		<-ticker.C
+		checkFreedPortsOnce()
+	}
+}
+
+func checkFreedPortsOnce() {
+	mu.Lock()
+	defer mu.Unlock()
+
+	pending := pendingPorts.Len()
+	remove := make([]*list.Element, 0, pending)
+	for elem := pendingPorts.Front(); elem != nil; elem = elem.Next() {
+		port := elem.Value.(int)
+		if used := isPortInUse(port); !used {
+			freePorts.PushBack(port)
+			remove = append(remove, elem)
+		}
+	}
+
+	retained := pending - len(remove)
+
+	if retained > 0 {
+		logf("WARN", "%d out of %d pending ports are still in use; something probably didn't wait around for the port to be closed!", pending-len(remove), pending)
+	}
+
+	if len(remove) == 0 {
+		return
+	}
+
+	for _, elem := range remove {
+		pendingPorts.Remove(elem)
+	}
+
+	condNotEmpty.Broadcast()
+}
+
+// adjustMaxBlocks avoids having the allocation ranges overlap the ephemeral
+// port range.
+func adjustMaxBlocks() int {
+	ephemeralPortMin, ephemeralPortMax, err := getEphemeralPortRange()
+	if err != nil {
+		logf("WARN", "ephemeral port range detection skipped due to error: %v", err)
```

Comment:
Initially I guess I didn't want all of the tests to fail because of this optimization going wrong, but now that you bring it up I'll have an error parsing the `sysctl` stuff hard-fail the tests so we actually find out about that in CI.

Reply:
Changed.